integrations/MULTI_GATEWAY_CLIENT.md

```diff
@@ -0,0 +1,175 @@
+
+These options are set on a `multi gateway client` level and apply to each request. If no options are specified the defaults as listed should be applied.
+
+### Request method
+
+There must be a method to allow IPFS data retrieval. The input for this method must be an IPFS url in these forms: `ipfs://<cid>` and `ipns://<cid>`.
+
+The data retrieval for a given CID must adhere to the configuration options.
+
+There must be an async way to get the data represented by that CID. While the `multi gateway client` can handle any CID data, in it's default settings all data is being verified. If `verify raw` is set to false then raw data is passed back as-is. CAR data is always verified.
+
+As an aside, a request is what the user puts in. For example:
+
+```cpp
+<client>::request("ipfs://bafyA");
+<client>::request("ipfs://bafyB");
+```
+
+Is 2 requests. These count at `max simultaneous cids` where the default is 5 maximum. If there are more then `max simultaneous cids` then those that don't get handled will be put on a queue to be handled as soon as a slot becomes available.
+
+Internally that CID is represented by N different CIDs (each block). Say `bafyA` consists of 100 blocks (simplified depiction):
+
+```
+bafyA
+| -- bafyA001
+| -- bafyA002
+| -- bafyA003
+| -- ...
+| -- bafyA100
+```
+
+Now the internal parts of the request method should download 10 of those block at the same time in an async way.
+
+The same would be true for `bafyB`.
+
+There still is the `max total gateways in use` of 25 by default. Which in this setup means that with 3 user requests you'd saturate the entire budget. In this case the novice approach would case an inbalance of gateway usage:
+
+```cpp
+<client>::request("ipfs://bafyA"); // handled by 10 gateways
+<client>::request("ipfs://bafyB"); // handled by 10 gateways
+<client>::request("ipfs://bafyC"); // handled by 5 gateways
+```
+
+The `multi gateway client` must detect that this N-th request is going to cause a gateway allocation inbalance and must rebalance the gateways. It should allocate an equal number of gateways to each request. If that equal number is below the `max total gateways in use` and there are requests with less then `concurrent requests` then the remainder of the budget should be allocated to one of those requests. A rebalanced example would look like this:
+
+```cpp
+<client>::request("ipfs://bafyA"); // handled by 9 gateways
+<client>::request("ipfs://bafyB"); // handled by 8 gateways
+<client>::request("ipfs://bafyC"); // handled by 8 gateways
+```
+
+This rebalancing must always happen, also in currently running requests.
+
+## Test fixtures
+
+N/A
+
+## Design rationale
+
+### User benefit
+
+Users, applications using IPFS, get a defined way to find gateways to use.
+
+### Compatibility
+
+N/A
+
+### Security
+
+N/A
```

Comment:
- Note that clients must verify received blocks before using them, and discard ones which do not match expected CID.
- If `ipns://` is to be supported, note if  / how  to handle DNSLinks